### DISCUSSION TITLE
Example NextJS multi language with typesafe-i18n package

### DISCUSSION THREAD

**Discussion by IRediTOTO** [vercel/next.js#28692] [Ideas]
Hi, i found a multi language that looks good.
https://github.com/ivanhofer/typesafe-i18n
Someone please make an example SSR with it? 
You will never miss a translation with typescript :)

**Comment by FDiskas** [comment]
It adds additional step to generate typings. I believe that typed translations possible to do  without as simple as `typeof myTranslationObject`

I successfully typed `next-i18next`

**Reply by IRediTOTO** [reply]
it have smaller size too, just ~1kb, next-translate is 4kb i think

**Reply by IRediTOTO** [reply]
btw, can you show me how to type with next-i18 ?

**Reply by FDiskas** [reply]
https://github.com/i18next/react-i18next/blob/master/example/react-typescript4.1/no-namespaces/%40types/react-i18next/index.d.ts

**Reply by FDiskas** [reply]
I actually made it more simple. I just import default language json file
```ts
import 'react-i18next';
import translation from './public/locales/en-AU/common.json';

declare module 'react-i18next' {
    type DefaultResources = { readonly translation: typeof translation };
    interface Resources extends DefaultResources {}
}
```

**Comment by belgattitude** [comment]
@IRediTOTO very nice indeed.

I like the lightweight approach. That said i18next is very mature solution, althought a bit bigger, so I think I'll stick on it for a while. 

As @FDiskas mentionned, you can type your translation keys with react-i18next too (and thus with next-i18next).

Some tip: dictionaries tend to grow fast, heavy and will be bundled. A nice thing is to do is to work with namespaces from the start. Requires a bit of organization at first (setting boundaries), but it avoids few hurdles later on.

As an example

```typescript
/**
 * Types augmentation for translation keys to allow to typecheck
 * and suggesting keys to the t function. In case it's too slow
 * you can opt out by commenting the following code.
 * @link https://react.i18next.com/latest/typescript
 */
import 'react-i18next';
import common from '../../public/locales/en/common.json';
import demo from '../../public/locales/en/demo.json';
import home from '../../public/locales/en/home.json';
import system from '../../public/locales/en/system.json';

declare module 'react-i18next' {
  interface CustomTypeOptions {
    defaultNS: 'common';
    resources: {
      common: typeof common;
      home: typeof home;
      demo: typeof demo;
      system: typeof system;
    };
  }
}
```

See [here](https://github.com/belgattitude/nextjs-monorepo-example/blob/main/apps/web-app/src/types.d/react-i18next.d.ts)

Then in my feature folders, I define which namespaces are enabled, an [example](https://github.com/belgattitude/nextjs-monorepo-example/blob/main/apps/web-app/src/features/home/home.config.ts).

Worth to say that when you deal with huge dictionaries it was really slow. Typescript 4.4.2 helped a lot with perf but autocompletion is not always working (typechecks yes).

**Comment by IRediTOTO** [comment]
Guys, i changed completely to typesafe-i18n :D 
Feel better, not much config, easy to use. Never miss a text. They will be soon create an example for Nextjs

**Reply by belgattitude** [reply]
Nice
